go-app-ussd_popi_user_data.js

```diff
@@ -694,14 +695,268 @@ go.app = function() {
+          if (identity.details.passport_no !== passport_no){
+            return self.states.create('state_incorrect_security_answers');
+          }
+        }
+        if(date_of_birth !== undefined){
+          if (identity.details.mom_dob !== date_of_birth){
+            return self.states.create('state_incorrect_security_answers');
+          }
+        }
+      return self.states.create('state_enter_new_phone_number');
+      });
+
+      self.add('state_incorrect_security_answers', function(name){
+        return new EndState(name, {
+          text: $("Sorry one or more of the answers you provided are incorrect. "+
+          "We are not able to change your mobile number. Please visit the clinic "+
+          "to register your new number."),
+          next: 'state_start'
+        });
+      });
+
+      self.add('state_enter_new_phone_number', function(name){
+        return new FreeText(name, {
+            question: $("Thank you. Please enter the new number you would like to use to receive messages from MomConnect."),
+            next: 'state_verify_new_number'
+        });
+      });
+
+      self.add('state_verify_new_number', function(name){
+        var new_number = self.im.user.get_answer("state_enter_new_phone_number");
+        return new MenuState(name, {
+          question: $("You have entered {{new_number}} as the new number you would like " +
+                      "to receive MomConnect messages on. Is this number correct?").context({new_number : new_number}),
+          choices: [
+              new Choice(
+                  'state_verify_new_number_in_database',
+                  $("Yes")),
+              new Choice('state_enter_new_phone_number', $("No - enter again"))
+          ]
+        });
+      });
+      self.add('state_verify_new_number_in_database', function(){
+        var msisdn = utils.normalize_msisdn(self.im.user.get_answer('state_enter_new_phone_number'), "27");
+        return is.get_or_create_identity({
+          msisdn: msisdn
+        }).then(function(identity){
+          return sbm.is_identity_subscribed(identity.id, [/^momconnect/, /^whatsapp/]);
+        }).then(function(subscribed){
+          if (subscribed){
+            return self.states.create("state_new_number_already_exists");
+          }
+          else{
+            return self.states.create("state_new_number_channel");
+          }
+        }
+      );
+      });
+
+      self.add("state_new_number_already_exists", function(name){
+        return new MenuState(name, {
+          question: $("Sorry the number you have entered is already associated with a "+
+                      "MomConnect account. Please try another number."),
+          choices: [
+              new Choice(
+                  'state_enter_new_phone_number',
+                  $("Try again")),
+              new Choice('state_exit', $("Exit"))
+          ]
+        });
+      });
+
+      self.add('state_new_number_channel', function(name) {
```

Comment:
I see we don't check if the new number is on whatsapp, is the thinking that we just switch them to SMS if the first message fails?

Reply:
Yes that is the thinking. Now that we have the automatic switch in place, I think it's better to rely on that, rather than rely on the number check API, which has rate-limiting.